File: fixtures/cpp/tests/calculator_test.cpp
#include "calculator.h"
#include "operations.h"
#include <cassert>
#include <cmath>
#include <iostream>

static const double EPSILON = 1e-9;

static void test_new_calculator() {
    Calculator calc;
    assert(std::abs(calc.lastResult()) < EPSILON);
}

static void test_add() {
    Calculator calc;
    double result = calc.add(2.0, 3.0);
    assert(std::abs(result - 5.0) < EPSILON);
}

static void test_subtract() {
    Calculator calc;
    double result = calc.subtract(5.0, 3.0);
    assert(std::abs(result - 2.0) < EPSILON);
}

static void test_multiply() {
    Calculator calc;
    double result = calc.multiply(4.0, 3.0);
    assert(std::abs(result - 12.0) < EPSILON);
}

static void test_divide() {
    Calculator calc;
    double result = calc.divide(10.0, 2.0);
    assert(std::abs(result - 5.0) < EPSILON);
}

static void test_divide_by_zero() {
    Calculator calc;
    bool caught = false;
    try {
        calc.divide(10.0, 0.0);
    } catch (const operations::DivisionByZeroError &) {
        caught = true;
    }
    assert(caught);
}

static void test_last_result() {
    Calculator calc;
    calc.add(1.0, 2.0);
    assert(std::abs(calc.lastResult() - 3.0) < EPSILON);
    calc.multiply(3.0, 4.0);
    assert(std::abs(calc.lastResult() - 12.0) < EPSILON);
}

static void test_operations_add() {
    assert(std::abs(operations::add(2.0, 3.0) - 5.0) < EPSILON);
}

static void test_operations_subtract() {
    assert(std::abs(operations::subtract(5.0, 3.0) - 2.0) < EPSILON);
}

static void test_operations_multiply() {
    assert(std::abs(operations::multiply(4.0, 3.0) - 12.0) < EPSILON);
}

static void test_operations_divide() {
    assert(std::abs(operations::divide(10.0, 2.0) - 5.0) < EPSILON);
}

int main() {
    test_new_calculator();
    test_add();
    test_subtract();
    test_multiply();
    test_divide();
    test_divide_by_zero();
    test_last_result();
    test_operations_add();
    test_operations_subtract();
    test_operations_multiply();
    test_operations_divide();

    std::cout << "All tests passed." << std::endl;
    return 0;
}

File: fixtures/rust/tests/calculator_test.rs
use calculator::Calculator;

#[test]
fn test_calculator_workflow() {
    let mut calc = Calculator::new();

    let sum = calc.add(10.0, 5.0);
    assert!((sum - 15.0).abs() < f64::EPSILON);

    let diff = calc.subtract(sum, 3.0);
    assert!((diff - 12.0).abs() < f64::EPSILON);

    let product = calc.multiply(diff, 2.0);
    assert!((product - 24.0).abs() < f64::EPSILON);

    let quotient = calc.divide(product, 4.0).unwrap();
    assert!((quotient - 6.0).abs() < f64::EPSILON);

    assert!((calc.last_result() - 6.0).abs() < f64::EPSILON);
}

#[test]
fn test_division_by_zero_returns_error() {
    let mut calc = Calculator::new();
    let result = calc.divide(10.0, 0.0);
    assert!(result.is_err());
}

#[test]
fn test_default_constructor() {
    let calc = Calculator::default();
    assert!((calc.last_result() - 0.0).abs() < f64::EPSILON);
}
